File: services/blockchain_service.py
from web3 import Web3
import json
import os
from flask import current_app
import logging

class BlockchainService:
    def __init__(self):
        # Set up logging
        logging.basicConfig(level=logging.INFO)
        self.logger = logging.getLogger(__name__)

        # Default settings for offline mode
        self.offline_mode = True
        self.blockchain_enabled = False
        self.w3 = None
        self.token_contract = None
        self.document_contract = None
        self.wallet_factory_contract = None
        self.token_tracker_contract = None
        self.token_address = None
        self.document_address = None
        self.wallet_factory_address = None
        self.token_tracker_address = None

        # Check if blockchain is enabled in config
        if not current_app.config.get('BLOCKCHAIN_ENABLED', False):
            self.logger.info("Blockchain is disabled in config, operating in offline mode")
            return

        # Initialize Web3 connection
        provider_url = current_app.config['BLOCKCHAIN_PROVIDER']

        # If blockchain provider is empty or None, operate in offline mode
        if not provider_url:
            self.logger.warning("No blockchain provider specified, operating in offline mode")
            return

        try:
            # Initialize Web3 connection
            self.w3 = Web3(Web3.HTTPProvider(provider_url))

            # Validate blockchain connection
            if not self.w3.is_connected():
                self.logger.error("Failed to connect to blockchain provider")
                return  # Stay in offline mode instead of raising exception

            # Connection successful - set flags
            self.offline_mode = False
            self.blockchain_enabled = True

            # Load contract ABIs
            abis_path = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'blockchain', 'abis')

            # Create abis directory if it doesn't exist
            if not os.path.exists(abis_path):
                os.makedirs(abis_path)

            # Load existing contract ABIs with error handling
            try:
                with open(os.path.join(abis_path, 'token_abi.json'), 'r') as token_file:
                    self.token_abi = json.load(token_file)
            except FileNotFoundError:
                self.logger.error("Token ABI file not found, using placeholder")
                self.token_abi = []  # Empty ABI as placeholder

            try:
                with open(os.path.join(abis_path, 'document_abi.json'), 'r') as document_file:
                    self.document_abi = json.load(document_file)
            except FileNotFoundError:
                self.logger.error("Document ABI file not found, using placeholder")
                self.document_abi = []  # Empty ABI as placeholder

            # Load new contract ABIs with error handling
            try:
                with open(os.path.join(abis_path, 'user_wallet_factory_abi.json'), 'r') as wallet_file:
                    self.wallet_factory_abi = json.load(wallet_file)
            except FileNotFoundError:
                self.logger.error("User Wallet Factory ABI file not found, using placeholder")
                self.wallet_factory_abi = []  # Empty ABI as placeholder

            try:
                with open(os.path.join(abis_path, 'user_token_tracker_abi.json'), 'r') as token_tracker_file:
                    self.token_tracker_abi = json.load(token_tracker_file)
            except FileNotFoundError:
                self.logger.error("User Token Tracker ABI file not found, using placeholder")
                self.token_tracker_abi = []  # Empty ABI as placeholder

            # Contract addresses from config
            self.token_address = current_app.config['TOKEN_CONTRACT_ADDRESS']
            self.document_address = current_app.config['DOCUMENT_CONTRACT_ADDRESS']
            self.wallet_factory_address = current_app.config.get('USER_WALLET_FACTORY_ADDRESS')
            self.token_tracker_address = current_app.config.get('USER_TOKEN_TRACKER_ADDRESS')

            # Initialize contract instances with proper error handling
            try:
                self.token_contract = self.w3.eth.contract(
                    address=self.token_address,
                    abi=self.token_abi
                )
            except Exception as e:
                self.logger.error(f"Failed to initialize token contract: {e}")
                self.token_contract = None

            try:
                self.document_contract = self.w3.eth.contract(
                    address=self.document_address,
                    abi=self.document_abi
                )
            except Exception as e:
                self.logger.error(f"Failed to initialize document contract: {e}")
                self.document_contract = None

            try:
                self.wallet_factory_contract = self.w3.eth.contract(
                    address=self.wallet_factory_address,
                    abi=self.wallet_factory_abi
                )
            except Exception as e:
                self.logger.error(f"Failed to initialize wallet factory contract: {e}")
                self.wallet_factory_contract = None

            try:
                self.token_tracker_contract = self.w3.eth.contract(
                    address=self.token_tracker_address,
                    abi=self.token_tracker_abi
                )
            except Exception as e:
                self.logger.error(f"Failed to initialize token tracker contract: {e}")
                self.token_tracker_contract = None

        except Exception as e:
            self.logger.error(f"Blockchain service initialization error: {e}")
            # Make sure we're in offline mode if there was an error
            self.w3 = None
            self.blockchain_enabled = False
            self.offline_mode = True

    def is_connected(self):
        """Check if blockchain service is connected and operational"""
        if self.w3 is None or not self.blockchain_enabled:
            return False

        try:
            return self.w3.is_connected()
        except:
            return False

    def get_account(self, index=0):
        """Get account from the connected node (for testing)"""
        if not self.is_connected():
            self.logger.warning("Blockchain not connected, returning fake address")
            return "0x0000000000000000000000000000000000000000"  # Fake address

        try:
            return self.w3.eth.accounts[index]
        except Exception as e:
            self.logger.error(f"Error getting blockchain account: {e}")
            return "0x0000000000000000000000000000000000000000"  # Fake address as fallback

    def create_user_wallet(self, user_address):
        """Create a new wallet for a user"""
        if not self.is_connected() or self.wallet_factory_contract is None:
            self.logger.warning("Blockchain not connected or wallet factory not initialized, returning mock receipt")
            return {"transactionHash": b"0x0", "status": 1}  # Mock receipt

        try:
            tx_hash = self.wallet_factory_contract.functions.createUserWallet(
                user_address
            ).transact({'from': self.get_account()})

            receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)
            self.logger.info(f"User wallet created for {user_address}")
            return receipt
        except Exception as e:
            self.logger.error(f"Error creating user wallet: {e}")
            return {"transactionHash": b"0x0", "status": 0}  # Mock failed receipt

    def track_token_earning(self, user_address, amount, reason):
        """Track token earning for a user"""
        if not self.is_connected() or self.token_tracker_contract is None:
            self.logger.warning("Blockchain not connected or token tracker not initialized, returning mock receipt")
            return {"transactionHash": b"0x0", "status": 1}  # Mock receipt

        try:
            tx_hash = self.token_tracker_contract.functions.earnTokens(
                user_address, amount, reason
            ).transact({'from': self.get_account()})

            receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)
            self.logger.info(f"Tokens earned by {user_address}: {amount} for {reason}")
            return receipt
        except Exception as e:
            self.logger.error(f"Error tracking token earning: {e}")
            return {"transactionHash": b"0x0", "status": 0}  # Mock failed receipt

    def get_user_token_balance(self, user_address):
        """Get user's token balance from blockchain"""
        if not self.is_connected() or self.token_tracker_contract is None:
            self.logger.warning("Blockchain not connected or token tracker not initialized, returning default balance")
            return 10  # Default token balance for development

        try:
            balance = self.token_tracker_contract.functions.getUserTokenBalance(
                user_address
            ).call()
            return balance
        except Exception as e:
            self.logger.error(f"Error getting user token balance: {e}")
            return 0

    def verify_document(self, file_hash):
        """Verify a document's existence on blockchain"""
        if not self.is_connected() or self.document_contract is None:
            self.logger.warning("Blockchain not connected or document contract not initialized, returning mock verification")
            return True  # Mock verification for development

        try:
            if not isinstance(file_hash, str):
                self.logger.warning(f"Invalid hash type: {type(file_hash)}")
                return False

            result = self.document_contract.functions.verifyDocument(file_hash).call()
            self.logger.info(f"Document verification result for {file_hash}: {result}")
            return result
        except Exception as e:
            self.logger.error(f"Document verification error: {e}")
            return False

    def mint_document_nft(self, file_hash, metadata, is_assignment, deadline, from_account):
        """Mint a new document NFT"""
        if not self.is_connected() or self.document_contract is None:
            self.logger.warning("Blockchain not connected or document contract not initialized, returning mock NFT data")
            return (
                {"transactionHash": b"0x0", "status": 1},  # Mock receipt
                "0x0000000000000000000000000000000000000000000000000000000000000000",  # Mock tx hash
                1  # Mock token ID
            )

        try:
            tx_hash = self.document_contract.functions.mintDocument(
                file_hash,
                metadata,
                is_assignment,
                deadline
            ).transact({'from': from_account})

            receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)

            # In a real implementation, we would parse the event logs to get the token ID
            # For simplicity in development, we'll just use a mock token ID
            token_id = 1  # This would normally come from event logs

            self.logger.info(f"Document minted with hash {file_hash}")
            return receipt, tx_hash.hex(), token_id
        except Exception as e:
            self.logger.error(f"Error minting document NFT: {e}")
            return (
                {"transactionHash": b"0x0", "status": 0},  # Mock failed receipt
                "0x0",  # Mock tx hash
                0  # Mock token ID (0 indicates failure)
            )

    def record_grade(self, token_id, grade_hash, from_account):
        """Record a grade for a document NFT"""
        if not self.is_connected() or self.document_contract is None:
            self.logger.warning("Blockchain not connected or document contract not initialized, returning mock grade data")
            return (
                {"transactionHash": b"0x0", "status": 1},  # Mock receipt
                "0x0000000000000000000000000000000000000000000000000000000000000000"  # Mock tx hash
            )

        try:
            tx_hash = self.document_contract.functions.recordGrade(
                token_id,
                grade_hash
            ).transact({'from': from_account})

            receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)
            self.logger.info(f"Grade recorded for token ID {token_id}")
            return receipt, tx_hash.hex()
        except Exception as e:
            self.logger.error(f"Error recording grade: {e}")
            return (
                {"transactionHash": b"0x0", "status": 0},  # Mock failed receipt
                "0x0"  # Mock tx hash
            )

    def get_grade(self, token_id):
        """Get the grade hash for a document NFT"""
        if not self.is_connected() or self.document_contract is None:
            self.logger.warning("Blockchain not connected or document contract not initialized, returning mock grade hash")
            return 12345  # Mock grade hash

        try:
            grade_hash = self.document_contract.functions.getGrade(token_id).call()
            return grade_hash
        except Exception as e:
            self.logger.error(f"Error getting grade: {e}")
            return 0

# Singleton pattern for blockchain service
blockchain_service = None

def get_blockchain_service():
    global blockchain_service
    if blockchain_service is None:
        blockchain_service = BlockchainService()
    return blockchain_service

File: moodle/local/blockchain/classes/blockchain_service.php
<?php
// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

namespace local_blockchain;

defined('MOODLE_INTERNAL') || die();

require_once($CFG->libdir . '/filelib.php');

class blockchain_service {
    private $web3;
    private $token_contract;
    private $document_contract;
    private $wallet_factory_contract;
    private $token_tracker_contract;
    private $offline_mode;
    private $blockchain_enabled;

    public function __construct() {
        global $CFG;
        
        $this->offline_mode = true;
        $this->blockchain_enabled = false;

        // Check if blockchain is enabled in config
        if (!get_config('local_blockchain', 'enabled')) {
            return;
        }

        try {
            // Initialize Web3 connection
            $provider_url = get_config('local_blockchain', 'provider_url');
            if (empty($provider_url)) {
                return;
            }

            // Initialize Web3 connection using PHP Web3 library
            $this->web3 = new \Web3\Web3($provider_url);

            // Load contract ABIs
            $abis_path = $CFG->dirroot . '/local/blockchain/contracts/abis';
            
            // Initialize contracts
            $this->initialize_contracts($abis_path);

            $this->offline_mode = false;
            $this->blockchain_enabled = true;

        } catch (\Exception $e) {
            debugging('Blockchain service initialization error: ' . $e->getMessage());
        }
    }

    private function initialize_contracts($abis_path) {
        // Load contract ABIs and addresses
        $token_abi = file_get_contents($abis_path . '/token_abi.json');
        $document_abi = file_get_contents($abis_path . '/document_abi.json');
        $wallet_factory_abi = file_get_contents($abis_path . '/user_wallet_factory_abi.json');
        $token_tracker_abi = file_get_contents($abis_path . '/user_token_tracker_abi.json');

        $token_address = get_config('local_blockchain', 'token_contract_address');
        $document_address = get_config('local_blockchain', 'document_contract_address');
        $wallet_factory_address = get_config('local_blockchain', 'wallet_factory_address');
        $token_tracker_address = get_config('local_blockchain', 'token_tracker_address');

        // Initialize contract instances
        $this->token_contract = $this->web3->eth->contract($token_abi, $token_address);
        $this->document_contract = $this->web3->eth->contract($document_abi, $document_address);
        $this->wallet_factory_contract = $this->web3->eth->contract($wallet_factory_abi, $wallet_factory_address);
        $this->token_tracker_contract = $this->web3->eth->contract($token_tracker_abi, $token_tracker_address);
    }

    public function verify_document($fileid) {
        global $DB;

        if ($this->offline_mode) {
            return true; // Simulate verification in offline mode
        }

        try {
            $document = $DB->get_record('local_blockchain_documents', ['fileid' => $fileid]);
            if (!$document) {
                return false;
            }

            // Verify document on blockchain
            $result = $this->document_contract->call('verifyDocument', [$document->hash]);
            
            if ($result) {
                $document->is_verified = 1;
                $document->timemodified = time();
                $DB->update_record('local_blockchain_documents', $document);
            }

            return $result;
        } catch (\Exception $e) {
            debugging('Document verification error: ' . $e->getMessage());
            return false;
        }
    }

    public function mint_document_nft($fileid, $metadata) {
        global $DB, $USER;

        if ($this->offline_mode) {
            return true; // Simulate NFT minting in offline mode
        }

        try {
            $document = $DB->get_record('local_blockchain_documents', ['fileid' => $fileid]);
            if (!$document) {
                return false;
            }

            // Mint NFT on blockchain
            $tx_hash = $this->document_contract->send('mintDocument', [
                $document->hash,
                $metadata,
                false, // is_assignment
                0 // deadline
            ], ['from' => $USER->blockchain_address]);

            if ($tx_hash) {
                $document->blockchain_tx = $tx_hash;
                $document->timemodified = time();
                $DB->update_record('local_blockchain_documents', $document);
                return true;
            }

            return false;
        } catch (\Exception $e) {
            debugging('NFT minting error: ' . $e->getMessage());
            return false;
        }
    }

    public function award_tokens($userid, $amount, $reason) {
        global $DB;

        if ($this->offline_mode) {
            // Record transaction locally in offline mode
            $transaction = (object)[
                'userid' => $userid,
                'amount' => $amount,
                'transaction_type' => 'reward',
                'reason' => $reason,
                'timecreated' => time()
            ];
            return $DB->insert_record('local_blockchain_tokens', $transaction);
        }

        try {
            $user = $DB->get_record('user', ['id' => $userid]);
            if (!$user || !$user->blockchain_address) {
                return false;
            }

            // Award tokens on blockchain
            $tx_hash = $this->token_tracker_contract->send('earnTokens', [
                $user->blockchain_address,
                $amount,
                $reason
            ]);

            if ($tx_hash) {
                // Record transaction
                $transaction = (object)[
                    'userid' => $userid,
                    'amount' => $amount,
                    'transaction_type' => 'reward',
                    'reason' => $reason,
                    'blockchain_tx' => $tx_hash,
                    'timecreated' => time()
                ];
                return $DB->insert_record('local_blockchain_tokens', $transaction);
            }

            return false;
        } catch (\Exception $e) {
            debugging('Token award error: ' . $e->getMessage());
            return false;
        }
    }

    public function check_integrity($fileid) {
        global $DB;

        if ($this->offline_mode) {
            return true; // Simulate integrity check in offline mode
        }

        try {
            $document = $DB->get_record('local_blockchain_documents', ['fileid' => $fileid]);
            if (!$document) {
                return false;
            }

            // Check document integrity on blockchain
            $result = $this->document_contract->call('checkIntegrity', [$document->hash]);
            
            if (!$result) {
                // Record integrity violation
                $violation = (object)[
                    'userid' => $document->userid,
                    'original_document_id' => $document->id,
                    'violation_type' => 'duplicate',
                    'description' => 'Document hash matches existing document on blockchain',
                    'status' => 'pending',
                    'timecreated' => time(),
                    'timemodified' => time()
                ];
                $DB->insert_record('local_blockchain_integrity', $violation);
            }

            return $result;
        } catch (\Exception $e) {
            debugging('Integrity check error: ' . $e->getMessage());
            return false;
        }
    }

    public function get_token_balance($userid) {
        global $DB;

        if ($this->offline_mode) {
            // Calculate balance from local transactions
            $rewards = $DB->get_records_sql(
                "SELECT SUM(amount) as total FROM {local_blockchain_tokens} 
                 WHERE userid = ? AND transaction_type = 'reward'",
                [$userid]
            );
            $spends = $DB->get_records_sql(
                "SELECT SUM(amount) as total FROM {local_blockchain_tokens} 
                 WHERE userid = ? AND transaction_type = 'spend'",
                [$userid]
            );
            
            $total_rewards = $rewards ? $rewards[0]->total : 0;
            $total_spends = $spends ? $spends[0]->total : 0;
            
            return $total_rewards - $total_spends;
        }

        try {
            $user = $DB->get_record('user', ['id' => $userid]);
            if (!$user || !$user->blockchain_address) {
                return 0;
            }

            // Get balance from blockchain
            return $this->token_tracker_contract->call('getUserTokenBalance', [$user->blockchain_address]);
        } catch (\Exception $e) {
            debugging('Token balance check error: ' . $e->getMessage());
            return 0;
        }
    }

    public function is_connected() {
        return !$this->offline_mode && $this->blockchain_enabled;
    }
} 